// @tableland/sdk@4.5.3-dev.0/dist/esm/validator/index.js
import { getBaseUrl, } from "../helpers/index.js";
import { getHealth } from "./health.js";
import { getVersion } from "./version.js";
import { getTable, } from "./tables.js";
import { getQuery, } from "./query.js";
import { getTransactionReceipt, pollTransactionReceipt, } from "./receipt.js";
export { ApiError } from "./client/index.js";
/**
 * Validator provides direct access to remote Validator REST APIs.
 */
export class Validator {
    /**
     * Create a Validator instance with the specified connection configuration.
     * @param config The connection configuration. This must include a baseUrl
     * string. If passing the config from a pre-existing Database instance, it
     * must have a non-null baseUrl key defined.
     */
    constructor(config = {}) {
        /* c8 ignore next 3 */
        if (config.baseUrl == null) {
            throw new Error("missing baseUrl information");
        }
        this.config = config;
    }
    /**
     * Create a new Validator instance that uses the default baseUrl for a given chain.
     * @param chainNameOrId The name or id of the chain to target.
     * @returns A Validator with a default baseUrl.
     */
    static forChain(chainNameOrId) {
        const baseUrl = getBaseUrl(chainNameOrId);
        return new Validator({ baseUrl });
    }
    /**
     * Get health status
     * @description Returns OK if the validator considers itself healthy
     */
    async health(opts = {}) {
        return await getHealth(this.config, opts);
    }
    /**
     * Get version information
     * @description Returns version information about the validator daemon
     */
    async version(opts = {}) {
        return await getVersion(this.config, opts);
    }
    /**
     * Get table information
     * @description Returns information about a single table, including schema information
     */
    async getTableById(params, opts = {}) {
        if (typeof params.chainId !== "number" ||
            typeof params.tableId !== "string") {
            throw new Error("cannot get table with invalid chain or table id");
        }
        return await getTable(this.config, params);
    }
    async queryByStatement(params, opts = {}) {
        return await getQuery(this.config, params, opts);
    }
    /**
     * Get transaction status
     * @description Returns the status of a given transaction receipt by hash
     */
    async receiptByTransactionHash(params, opts = {}) {
        return await getTransactionReceipt(this.config, params, opts);
    }
    /**
     * Wait for transaction status
     * @description Polls for the status of a given transaction receipt by hash until
     */
    async pollForReceiptByTransactionHash(params, opts = {}) {
        return await pollTransactionReceipt(this.config, params, opts);
    }
}
